src/target/target_flash.c

```diff
@@ -0,0 +1,271 @@
+		ret = f->prepare(f);
+
+	if (ret == 0)
+		f->ready = true;
+
+	return ret;
+}
+
+static int flash_done(target_flash_s *f)
+{
+	if (!f->ready)
+		return 0;
+
+	int ret = 0;
+	if (f->done)
+		ret = f->done(f);
+
+	if (f->buf) {
+		free(f->buf);
+		f->buf = NULL;
+	}
+
+	f->ready = false;
+
+	return ret;
+}
+
+int target_flash_erase(target *t, target_addr_t addr, size_t len)
+{
+	if (target_enter_flash_mode(t) != 0)
+		return 1;
+
+	int ret = 0;
+	while (len) {
+		target_flash_s *f = target_flash_for_addr(t, addr);
+		if (!f) {
+			DEBUG_WARN("Requested address is outside the valid range 0x%06" PRIx32 "\n", addr);
+			return 1;
+		}
+
+		/* terminate flash operations if we're not in the same target flash */
+		for (target_flash_s *target_f = t->flash; target_f; target_f = target_f->next)
+			if (target_f != f)
+				ret |= flash_done(target_f);
+
+		const target_addr_t local_start_addr = addr & ~(f->blocksize - 1U);
+		const target_addr_t local_end_addr = local_start_addr + f->blocksize;
+
+		if(flash_prepare(f) != 0)
+			return -1;
+
+		ret |= f->erase(f, local_start_addr, f->blocksize);
+
+		len -= MIN(local_end_addr - addr, len);
+		addr = local_end_addr;
+
+		/* Issue flash done on last operation */
+		if (len == 0)
+			ret |= flash_done(f);
+	}
+	return ret;
+}
+
+int target_flash_write(target *t, target_addr_t dest, const void *src, size_t len)
+{
+	if (target_enter_flash_mode(t) != 0)
+		return 1;
+
+	int ret = 0;
+	while (len) {
+		target_flash_s *f = target_flash_for_addr(t, dest);
+		if (!f)
+			return 1;
+
+		/* terminate flash operations if we're not in the same target flash */
+		for (target_flash_s *target_f = t->flash; target_f; target_f = target_f->next) {
+			if (target_f != f) {
+				ret |= flash_buffered_flush(target_f);
+				ret |= flash_done(target_f);
+			}
+		}
+
+		const target_addr_t local_end_addr = MIN(dest + len, f->start + f->length);
+		const target_addr_t local_length = local_end_addr - dest;
+
+		ret |= flash_buffered_write(f, dest, src, local_length);
+
+		dest = local_end_addr;
+		src += local_length;
+		len -= local_length;
+
+		/* Flush operations if we reached the end of Flash */
+		if (dest == f->start + f->length) {
+			ret |= flash_buffered_flush(f);
+			ret |= flash_done(f);
+		}
+	}
+	return ret;
+}
+
+int target_flash_terminate(target *t)
+{
+	if (!t->flash_mode)
+		return -1;
+
+	int ret = 0;
+	for (target_flash_s *f = t->flash; f; f = f->next) {
+		ret |= flash_buffered_flush(f);
+		ret |= flash_done(f);
+	}
+
+	target_exit_flash_mode(t);
+	return ret;
+}
+
+static int flash_buffered_write(target_flash_s *f, target_addr_t dest, const void *src, size_t len)
+{
+	if (f->buf == NULL) {
+		/* Allocate buffer */
+		f->buf = malloc(f->writebufsize);
+		if (!f->buf) { /* malloc failed: heap exhaustion */
+			DEBUG_WARN("malloc: failed in %s\n", __func__);
+			return -1;
+		}
+		f->buf_addr_base = UINT32_MAX;
+		f->buf_addr_low = UINT32_MAX;
+		f->buf_addr_high = 0;
+	}
+
+	int ret = 0;
+	while (len) {
```

Comment:
Is it possible to pivot this into a for loop?

Reply:
here It might be possible, but I don't see the immediate answear, so I will leave it be